### PR TITLE
`PluginManager#whichPlugin` does not work for `JenkinsRule`

### PR DESCRIPTION
**Pull request by basil**
Observed in the context of a proprietary plugin: `PluginManager#whichPlugin` does not work for `JenkinsRule` (i.e., plugins in the test classpath). The new unit test fails before the changes to `src/main` and passes after. I created a new test class because the existing `PluginManagerTest` uses `PluginManagerUtil.newJenkinsRule()` rather than `new JenkinsRule()` and therefore can't set up the expected value correctly.

### Proposed changelog entries

Developer: Support usage of `PluginManager#whichPlugin` in `JenkinsRule`-based tests.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6982"></a>

### CODE REVIEW

File: core/src/main/java/hudson/PluginManager.java
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java:1344]: Out of curiosity, what examples did you see which were directories? I could imagine `some-plugin/target/classes` but only if you had a `*-SNAPSHOT` dependency on the plugin. In that case you would ideally actually load `some-plugin/target/classes/META-INF/MANIFEST.MF`, to allow tests to pass in developer mode prior to commit, but that is definitely a corner case.

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/PluginManager.java
Comment (basil) [reply core/src/main/java/hudson/PluginManager.java:1344]: > Out of curiosity, what examples did you see which were directories?

```
   3.192 [id=35]        WARNING hudson.PluginManager#whichPlugin: could not inspect file:src/jenkinsci/jenkins/test/target/test-classes/
java.io.FileNotFoundException: src/jenkinsci/jenkins/test/target/test-classes (Is a directory)
        at java.base/java.io.RandomAccessFile.open0(Native Method)
        at java.base/java.io.RandomAccessFile.open(RandomAccessFile.java:345)
        at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:259)
        at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:214)
        at java.base/java.util.zip.ZipFile$Source.<init>(ZipFile.java:1305)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1271)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:733)
        at java.base/java.util.zip.ZipFile$CleanableResource.get(ZipFile.java:850)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:248)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:177)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:350)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:321)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:287)
        at hudson.PluginManager.whichPlugin(PluginManager.java:1344)
        at hudson.diagnosis.NullIdDescriptorMonitor.verify(NullIdDescriptorMonitor.java:73)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:109)
        at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:185)
        at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:305)
        at jenkins.model.Jenkins$5.runTask(Jenkins.java:1160)
        at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:222)
        at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:121)
        at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:70)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

Comment (jglick) [reply core/src/main/java/hudson/PluginManager.java:1344]: Oh right, if called on something which is indeed not a plugin.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
interesting I thought `getPlugin()` didn't work in `JenkinsRule` (https://issues.jenkins.io/browse/JENKINS-48885)

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by daniel-beck — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by jglick** [comment]
> I thought `getPlugin()` didn't work in `JenkinsRule`

https://github.com/jenkinsci/jenkins-test-harness/pull/92 sounds similar but is not.

File: core/src/main/java/hudson/PluginManager.java
Comment (jglick) [reply core/src/main/java/hudson/PluginManager.java:1344]: (But confirming my suspicion—this is similar to a snapshot dependency, but for unpackaged class files within the same Maven module.)

**Pull request merged**
